host/src/gatt.rs

```diff
@@ -25,152 +26,304 @@ use crate::types::gatt_traits::GattValue;
+            &mut self.pdu,
+            handle,
+            &self.connection,
+            self.server,
+            self.tx_pool,
+            result,
+        )
+    }
+
+    /// Process and respond to event.
+    pub async fn reply(mut self, result: Result<(), AttErrorCode>) -> Result<(), Error> {
+        let handle = self.handle();
+        reply(
+            &mut self.pdu,
+            handle,
+            &self.connection,
+            self.server,
+            self.tx_pool,
+            result,
+        )
+        .await
+    }
+}
+
+impl Drop for ReadEvent<'_, '_> {
+    fn drop(&mut self) {
+        let handle = self.handle();
+        let _ = try_reply(
+            &mut self.pdu,
+            handle,
+            &self.connection,
+            self.server,
+            self.tx_pool,
+            Ok(()),
+        );
+    }
+}
+
+/// An event returned while processing GATT requests.
+pub struct WriteEvent<'d, 'server> {
+    /// Characteristic handle that was written.
+    value_handle: u16,
+    pdu: Option<Pdu<'d>>,
+    connection: Connection<'d>,
+    tx_pool: &'d dyn GlobalPacketPool<'d>,
+    server: &'server dyn DynamicAttributeServer,
+}
+
+impl<'d, 'server> WriteEvent<'d, 'server> {
+    /// Characteristic handle that was read
+    pub fn handle(&self) -> u16 {
+        self.value_handle
+    }
+
+    /// Characteristic data that was written
+    pub fn data(&self) -> &[u8] {
+        // Note: write event data is always at offset 3, right?
+        &self.pdu.as_ref().unwrap().as_ref()[3..]
+    }
+
+    /// Process and respond to event.
+    pub fn try_reply(mut self, result: Result<(), AttErrorCode>) -> Result<(), Error> {
+        let handle = self.handle();
+        try_reply(
+            &mut self.pdu,
+            handle,
+            &self.connection,
+            self.server,
+            self.tx_pool,
+            result,
+        )
+    }
+
+    /// Process and respond to event.
+    pub async fn reply(mut self, result: Result<(), AttErrorCode>) -> Result<(), Error> {
+        let handle = self.handle();
+        reply(
+            &mut self.pdu,
+            handle,
+            &self.connection,
+            self.server,
+            self.tx_pool,
+            result,
+        )
+        .await
+    }
+}
+
+impl Drop for WriteEvent<'_, '_> {
+    fn drop(&mut self) {
+        let handle = self.handle();
+        let _ = try_reply(
+            &mut self.pdu,
+            handle,
+            &self.connection,
+            self.server,
+            self.tx_pool,
+            Ok(()),
+        );
+    }
+}
+
+async fn reply<'d>(
+    pdu: &mut Option<Pdu<'d>>,
+    handle: u16,
+    connection: &Connection<'d>,
+    server: &dyn DynamicAttributeServer,
+    tx_pool: &'d dyn GlobalPacketPool<'d>,
+    result: Result<(), AttErrorCode>,
+) -> Result<(), Error> {
+    if let Some(pdu) = pdu.take() {
+        match result {
+            Ok(_) => {
+                let att = unwrap!(AttReq::decode(pdu.as_ref()));
```

Comment:
Is there any potential for `decode` to fail in this instance? If not, is it worth annotating this just to make it clear?

Reply:
It's already been decoded when dispatching from the host runner, so yeah. I'll add a comment :+1: